Sources/MongoSwift/BSON/Document+Collection.swift

```diff
@@ -0,0 +1,49 @@
+import Foundation
+
+/// An extension of `Document` to make it conform to the `Collection` protocol.
+/// This gives guarantees on non-destructive iteration, and offers an indexed
+/// ordering to the key-value pairs in the document.
+extension Document: Collection {
+    /// Returns the start index of the Document.
+    public var startIndex: Int {
+        return 0
+    }
+
+    /// Returns the end index of the Document.
+    public var endIndex: Int {
+        return self.count
+    }
+
+    private func failIndexCheck(_ i: Int) {
+        let invalidIndexMsg = "Index \(i) is invalid"
+        if self.count == 0 {
+            preconditionFailure(invalidIndexMsg)
+        }
+        precondition(self.startIndex ... self.endIndex - 1 ~= i, invalidIndexMsg)
+    }
+
+    /// Returns the index after the given index for this Document.
+    public func index(after i: Int) -> Int {
+        // Index must be a valid one, meaning it must exist somewhere in self.keys.
+        failIndexCheck(i)
+        return i + 1
+    }
+
+    /// Allows access to a `KeyValuePair` from the `Document`, given the position of the desired `KeyValuePair` held
+    /// within. This method does not guarantee constant-time (O(1)) access.
+    public subscript(position: Int) -> Document.KeyValuePair {
+        // TODO: This method _should_ guarantee constant-time O(1) access, and it is possible to make it do so. This
+        // criticism also applies to key-based subscripting via `String`.
+        // See SWIFT-250.
+        failIndexCheck(position)
+        return self.keyValuePairs[position]
```

Comment:
could we use `DocumentIterator.subsequence` here instead? or maybe write a variant of that method that uses the same approach to just gets one `KeyValuePair` value given its index?  ideally, we should just advance the iterator to the desired position and return what's there. currently, this requires us to always traverse the whole document and copy every key-value pair over into Swift, but then we just end up only looking at one of them.

Reply:
Yes, I've opted for:
```swift
// Because of our failIndexCheck precondition, this call is guaranteed to provide a non-nil result.
return DocumentIterator.subsequence(of: self, startIndex: position, endIndex: position + 1).first!
```

However, this looks kind of gnarly, the really simple approach one thinks of first would have been:
```swift
let iterator = makeIterator()
if position == 0 { return iterator.next()! }
for _ in 0..<position {
    iterator.advance()
}
// Because of our failIndexCheck precondition, this call is guaranteed to provide a non-nil result.
return iterator.next()!
```
Which is not much prettier. I've opted for the `subsequence` approach because it seems to be a lot easier to understand as being basically a way of getting a single-element subsequence.

Thoughts?

Reply:
let's just stick with subsequence, I think it is pretty readable and we may as well not introduce more logic if this is going to change before long anyway.